File: 31-next-permutation/31-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if(nums.size()<=1) return;
        int i=nums.size()-2;
        while(i>=0&&nums[i]>=nums[i+1]) i--;
        int j=nums.size()-1;
        while(i>=0&&j>i&&nums[i]>=nums[j]) j--;
        if(i>=0){
            swap(nums[i],nums[j]);
        }
        reverse(nums.begin()+i+1,nums.end());
    }
};

File: 31-next-permutation/31-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int indx1=nums.length-2;
        while(indx1>=0&&nums[indx1]>=nums[indx1+1]) {
            indx1--;
        }
        if(indx1>=0){
            int indx2=nums.length-1;
            while(indx2>indx1&&nums[indx2]<=nums[indx1]) indx2--;
            nums[indx1]=nums[indx1]+nums[indx2];
            nums[indx2]=nums[indx1]-nums[indx2];
            nums[indx1]=nums[indx1]-nums[indx2];
        }
        reverse(nums,indx1+1);
    }
    void reverse(int[] nums,int i){
        int s=i;
        int e=nums.length-1;
        while(s<e){
            nums[s]=nums[s]+nums[e];
            nums[e]=nums[s]-nums[e];
            nums[s]=nums[s]-nums[e];
            s++;e--;
        }
    }
}